puffin/src/scope_details.rs

```diff
@@ -0,0 +1,212 @@
+        let mut new_scopes = HashSet::new();
+        for scope_detail in scopes {
+            let new_scope_id = fetch_add_scope_id();
+
+            self.insert(new_scope_id, ScopeDetailsOwned::from(scope_detail));
+
+            new_scopes.insert(new_scope_id);
+        }
+
+        new_scopes
+    }
+
+    /// Fetches all registered scopes and their ids.
+    /// Useful for fetching scope id by a static scope name.
+    pub fn scopes_by_name<T>(
+        &self,
+        mut existing_scopes: impl FnMut(&std::collections::HashMap<String, ScopeId>) -> T,
+    ) -> T {
+        existing_scopes(&self.0.read().string_to_scope_id)
+    }
+
+    /// Fetches all registered scopes.
+    /// Useful for fetching scope details by a scope id.
+    pub fn scopes_by_id<T>(
+        &self,
+        mut existing_scopes: impl FnMut(&std::collections::HashMap<ScopeId, ScopeDetailsOwned>) -> T,
+    ) -> T {
+        existing_scopes(&self.0.read().scope_id_to_details)
+    }
+}
+
+/// Scope details that are registered by the macros.
+/// This is only used internally and should not be exposed.
+#[derive(Debug, Clone, Copy, PartialEq, Hash, PartialOrd, Ord, Eq)]
+pub struct ScopeDetailsRef {
+    /// Identifier for the scope being registered.
+    pub scope_id: ScopeId,
+    // Custom provided static id that identifiers a scope in a function.
+    pub scope_name: &'static str,
+    /// Scope name, or function name (previously called "id")
+    pub raw_function_name: &'static str,
+    /// Path to the file containing the profiling macro
+    pub file: &'static str,
+    /// The line number containing the profiling
+    pub line_nr: u32,
+}
+
+/// Custom scope details that can be registered by external users.
+/// Instantiate this type once for each custom scope that you record manually.
+/// Custom scopes can be registered via `GlobalProfiler::scope_details().insert_scopes()`.
+// This type provides slightly more convenient api for external users.
+#[derive(Debug, Clone, PartialEq, Hash, PartialOrd, Ord, Eq)]
+pub struct CustomScopeDetails {
+    /// Unique identifier for this scope.
+    pub scope_name: Cow<'static, str>,
+    /// Scope name, or function name (previously called "id")
+    pub function_name: Cow<'static, str>,
+    /// Path to the file containing the profiling macro
+    /// Can be empty if unused.
+    pub file_name: Cow<'static, str>,
+    /// The line number containing the profiling.
+    /// Can be 0 if unused.
+    pub line_nr: u32,
+}
+
+impl CustomScopeDetails {
+    /// Create a new custom scope with a unique name.
+    pub fn new(scope_name: impl Into<Cow<'static, str>>) -> Self {
+        Self {
+            scope_name: scope_name.into(),
+            function_name: Default::default(),
+            file_name: Default::default(),
+            line_nr: Default::default(),
+        }
+    }
+
+    pub fn with_function_name(mut self, name: impl Into<Cow<'static, str>>) -> Self {
```

Comment:
It is very important that all public small functions like these are marked `#[inline]`